### PR TITLE
Added MultiMeshInstance2D node for using MultiMesh in 2D

### LINKED ISSUE #28304: SurfaceTool (and Mesh) are not included in disable_3d=yes build
**Godot version:**

3.1.1.devel cab492147 (compiled editor and templates)

**OS/device including version:**

Kubuntu 19.04

**Issue description:**

it runs fine with using `SurfaceTool` in script.
but exported game crashes on both debug and release export.

```
$ ./parse_error_surface_tool.x86_64 
OpenGL ES 3.0 Renderer: GeForce GTX 1070 Ti/PCIe/SSE2
SCRIPT ERROR: GDScript::load_byte_code: Parse Error: Identifier 'SurfaceTool' is not declared in the current scope.
   At: res://Node2D.gdc:5.
ERROR: load_byte_code: Method/Function Failed, returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:785.
ERROR: load: Condition ' err != OK ' is true. returned: RES()
   At: modules/gdscript/gdscript.cpp:2168.
ERROR: _load: Failed loading resource: res://Node2D.gdc
   At: core/io/resource_loader.cpp:285.
ERROR: poll: res://Node2D.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Node2D.gd
   At: scene/resources/resource_format_text.cpp:440.
ERROR: load: Condition ' err != OK ' is true. returned: RES()
   At: core/io/resource_loader.cpp:208.
ERROR: _load: Failed loading resource: res://Node2D.tscn
   At: core/io/resource_loader.cpp:285.
ERROR: start: Failed loading scene: res://Node2D.tscn
   At: main/main.cpp:1739.
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x43f60) [0x7fa14cb44f60] (??:0)
[2] ./parse_error_surface_tool.x86_64() [0x1047c5f] (??:?)
[3] ./parse_error_surface_tool.x86_64() [0xb33abc] (??:?)
[4] ./parse_error_surface_tool.x86_64() [0x59e402] (??:?)
[5] ./parse_error_surface_tool.x86_64(main+0x84) [0x578f24] (??:?)
[6] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7fa14cb27b6b] (??:0)
[7] ./parse_error_surface_tool.x86_64() [0x57a39a] (??:?)
-- END OF BACKTRACE --
Aborted (core dumped)
```

**Steps to reproduce:**
```gdscript
extends Node2D

func _ready():
	var surface = SurfaceTool.new()
	pass
```

**Minimal reproduction project:**

sample project : [parse_error_surface_tool.zip](https://github.com/godotengine/godot/files/3103567/parse_error_surface_tool.zip)

exported : [parse_error_surface_tool.x86_64.zip](https://github.com/godotengine/godot/files/3103573/parse_error_surface_tool.x86_64.zip)

### LINKED ISSUE #29407: MultiAtlasSprite? Or other performant way to render many 2d sprites
**Godot version:**

3.1.1

I'm going to have a (simple) custom physics simulation in C++ for a bunch (thousands) of 2D bullets, and couldn't find a nice way in the current API to render them. Each bullet may have a different texture, though most will have one of a few very common textures.

I feel like the best way to render these would be instanced rendering using a texture atlas for the bullets, but I don't see how I could do this using the Godot API. It would be nice to have something like MultiMesh for 2d textured rectangles, except each instance can specify a different part of an atlas to pull from, as well as specify a different rectangle size.

Additionally, it would be useful to be able to specify a parameter for all instances at once, using a PoolArray (e.g. use a PoolVector2Array to specify all positions). All bullets will always be moving, so I will need to update all of their positions each frame (Other parameters may change less frequently). I don't want to have to incur the cost of a function call for each bullet, when setting instance parameters through the bindings produced by godot-cpp.

Methods in the current API I have considered, and what I think the issues with them are:
* Sprite for each bullet - incurs the cost of a node for every bullet, and it's commonly advised not to use Sprites if you have many similar things to render. Doesn't allow batch updates.
* Texture rectangle draw commands on a CanvasItem / through VisualServer - Doesn't require a node for each bullet, but still requires a separate draw call for each, and may or may not require other context switching (I don't know how it's implemented). Doesn't allow batch updates. I'll probably use this if it's fast enough and there's nothing better.
* MultiMesh - I would have to work on my 2d game in 3d space, and it doesn't allow batch updates. Also not sure how I would make it work with textures, but I guess I could put UVs in the custom data and add a shader with a texture uniform? Probably my most likely solution if there isn't something better for 2d, and if draw commands are too slow.

For context, I'm doing a custom simulation because I want the simulation to be deterministic across platforms, hardware, etc. To do that, I'm using fixed precision fractions (implemented with integer math).

### PR DESCRIPTION
**Pull request by clayjohn**
Also fixes bug where not using ``instance_custom`` would crash shader and bug where ``instance_color`` defaulted to black in GLES2.

I will make a corresponding PR in godot-design for the new icon. 

Fixes: https://github.com/godotengine/godot/issues/29407
Bugsquad edit: Fixes #28304 incidentally.

### CODE REVIEW

File: scene/2d/multimesh_instance_2d.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
+/*************************************************************************/
+/*  multimesh_instance.cpp                                               */
```
Comment (akien-mga) [inline scene/2d/multimesh_instance_2d.cpp]: Should be `multimesh_instance_2d.cpp` (note that the trailing `*/` will become unaligned, so you need to remove 3 spaces too).

File: scene/2d/multimesh_instance_2d.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	normal_map = p_texture;
+	update();
+}
+
+Ref<Texture> MultiMeshInstance2D::get_normal_map() const {
+
+	return normal_map;
+}
+
+Ref<Texture> MultiMeshInstance2D::get_texture() const {
```
Comment (akien-mga) [inline scene/2d/multimesh_instance_2d.cpp]: Would be good to move `get_texture` below `set_texture`, we usually group setter and getter together.

File: scene/2d/multimesh_instance_2d.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*************************************************************************/
+/*  multimesh_instance.h                                                 */
```
Comment (akien-mga) [inline scene/2d/multimesh_instance_2d.h]: Same here.

**Comment by akien-mga** [comment]
This will likely fail to compile with `disable_3d=yes` (the updated CI from #29405 should tell us on the next update), as `MultiMesh` is not registered in `scene/register_scene_types.cpp` if `_3D_DISABLED` is defined. So you'd have to fix that.

**Comment by akien-mga** [comment]
I guess to expose `MultiMesh` to `_3D_DISABLED` builds, you might need to expose `Mesh` too, and could thus also fix #28304.

**Comment by clayjohn** [comment]
All your comments should be addressed now.

I moved up ``Mesh``, ``MultiMesh``, ``ArrayMesh``, ``SurfaceTool``, and ``MeshDataTool``. All can be used in 2D so they should be included with 3D_disabled.

I opted not to include the various ``PrimitiveMeshes``, but we may want to include that in a later build. especially ``QuadMesh`` which can be very useful in 2D.

**Comment by akien-mga** [comment]
Thanks!

**Comment by raymoo** [comment]
Is there a usage example available? I'm trying to use MultiMesh in 2D in the stable version and am not sure if I am modifying the instances incorrectly (https://paste.debian.net/1086832/ https://paste.debian.net/1086833/)

**Comment by clayjohn** [comment]
@raymoo It all looks fine. The only thing that looks off is the size of your QuadMesh. What issue are you having exactly?

**Comment by raymoo** [comment]
In case anyone reading this issue is wondering, my issue is that I was sharing the multimesh with all my nodes, because I thought _process would be called immediately before each one is rendered. Many of my bullet clouds are empty, so nothing was being rendered. I solved it by not sharing the multimesh.

**Pull request merged**